Add `robot_step_begin()` and `robot_step_end()` API functions for controller optimization (#4107)
cyberbotics/webots: docs/reference/robot.md

    Robot();
    virtual ~Robot();
    virtual int step(int duration);

>     virtual int stepBegin(int duration);
>     virtual int stepEnd();

Reviewer: ```suggestion
    int stepBegin(int duration);
    int stepEnd();
```

During this time, the controller is idle, waiting for Webots to complete its simulation step.
On the other hand, prior to starting a new step, Webots waits for all the controllers to send their `wb_robot_step` messages which may induce some idle waiting time in Webots if a controller doesn't send quickly enough its `wb_robot_step` message because it is busy with some computation.
If the two computational processes (Webots and controller) are slow, it may be interesting to parallelize them.
`wb_robot_step_begin` and `wb_robot_step_end` allow you to achieve such an implementation.
They correspond to a split version of `wb_robot_step`, with the particularity that the code written between the two function calls is executed in parallel with the Webots simulation step.
`wb_robot_step_begin` and `wb_robot_step_end` both return -1 if the simulation is terminated.

> Note that some functions cannot be implemented between `wb_robot_step_begin` and `wb_robot_step_end`, as they require immediate response from Webots using a request-response pattern. 

Reviewer: ```suggestion
Note that some Webots API functions cannot be called between `wb_robot_step_begin` and `wb_robot_step_end`, as they require immediate response from Webots using a request-response pattern. 
```

This includes some [Supervisor API](supervisor.md) functions, like `wb_supervisor_node_get_field` and `wb_supervisor_field_get_sf_rotation`, but also some Robot API functions, like `wb_robot_get_urdf`. 

> Webots warns you in case you implement one of these functions between `wb_robot_step_begin` and `wb_robot_step_end`.

Reviewer: ```suggestion
Webots will warn you in case you call one of these functions between `wb_robot_step_begin` and `wb_robot_step_end`.
```

> You can simply write them before `wb_robot_step_begin` or after `wb_robot_step_end`.

Reviewer: ```suggestion
You can simply call them before `wb_robot_step_begin` or after `wb_robot_step_end`.
```

> If you still want to parallelize some of these functions with the simulation step, you can use the tracking options from the supervisor. 

Reviewer: ```suggestion
However, some of these functions can be called between `wb_robot_step_begin` and `wb_robot_step_end` if you enable the supervisor tracking feature. 
```

> `wb_supervisor_field_enable_sf_tracking`, `wb_supervisor_node_enable_pose_tracking` and `wb_supervisor_node_enable_contact_point_tracking` force Webots to stream information to the controller. 

Reviewer: ```suggestion
`wb_supervisor_field_enable_sf_tracking`, `wb_supervisor_node_enable_pose_tracking` and `wb_supervisor_node_enable_contact_point_tracking` force Webots to continuously stream the requested information to the controller.
```

> Using tracking, some supervisor methods get their value locally and can be written between `wb_robot_step_begin` and `wb_robot_step_end`.

Reviewer: ```suggestion
By enabling the tracking, the corresponding supervisor functions can be called between `wb_robot_step_begin` and `wb_robot_step_end`, because their value will be queried to Webots during `wb_robot_step_begin` and received during `wb_robot_step_end`.
```

      // leave the loop when the simulation is over
    } while (stepEnd() != -1);
  }

private:
  int timeStep;
  DistanceSensor *distanceSensor;
  Led *led;
}

// main C++ program
int main() {
  MyController *controller = new MyController();
  controller->run();
  delete controller;
  return 0;
}

```

%tab-end

%tab "Python"

```python
from controller import Robot

class MyController(Robot):
    def __init__(self):
        super(MyController, self).__init__()
        self.timeStep = 32  # set the control time step

        # get device tags
        self.distanceSensor = self.getDistanceSensor('my_distance_sensor')
        self.led = self.getLed('my_led')
        self.distanceSensor.enable(timeStep)  # enable sensors to read data from them

    def run(self):
        # with a parallelized control loop, it may be necessary to run an initial step to initialize sensor values
        self.step(self.timeStep)

        # main control loop
        while True:
            # begin simulation step computation: send command values to Webots for update
            # leave the loop when the simulation is over
            if self.stepBegin(self.timeStep) == -1:

>               break

Reviewer: Python requires 4 spaces indent:
```suggestion
                break
```

            
            # the following code (until self.step_end) is executed in parallel with the background simulation step

            val = self.distanceSensor.getValue()  # Read and process sensor data

            # Intensive computation could take place here

            self.led.set(1)                       # Send actuator commands
            
            # end simulation step computation: retrieve new sensor values from Webots
            # leave the loop when the simulation is over
            if self.stepEnd() == -1:

>               break

Reviewer: ```suggestion
                break
```
